src/langchain_google_memorystore_redis/doc_saver.py

```diff
@@ -0,0 +1,97 @@
+from langchain_core.documents.base import Document
+
+
+class MemorystoreDocumentSaver:
+    """Document Saver for Cloud Memorystore for Redis database."""
+
+    def __init__(
+        self,
+        client: redis.Redis,
+        content_field: str,
+        metadata_fields: Optional[Set[str]] = None,
+        key_prefix: Optional[str] = None,
+    ):
+        """Initializes the Document Saver for Memorystore for Redis.
+
+        Args:
+            client: A redis.Redis client object.
+            content_field: The field of the hash that Redis uses to store the
+                page_content of the Document.
+            metadata_fields: The metadata fields of the Document that will be
+                stored in the Redis. If None, Redis stores all metadata fields.
+            key_prefix: A prefix for the key used to store history for this
+                session.
+        """
+
+        self._redis = client
+        self._content_field = content_field
+        self._metadata_fields = metadata_fields
+        self._key_prefix = key_prefix if key_prefix else ""
+
+    def add_documents(
+        self,
+        documents: Sequence[Document],
+        ids: Optional[Sequence[str]] = None,
+    ) -> None:
+        """Save a list of Documents to Redis.
+
+        Args:
+            documents: A List of Documents.
+            ids: The keys in Redis used to store the Documents. If specified,
+                the length of the IDs must be the same as Documents. If not
+                specified, random UUIDs with prefix are generated to store
+                each Document.
+        """
+        doc_ids = ids
+        if not doc_ids:
+            doc_ids = [self._key_prefix + str(uuid.uuid4()) for _ in documents]
+        if len(documents) != len(doc_ids):
+            raise ValueError("The length of documents must match the length of the IDs")
+
+        for i, doc in enumerate(documents):
+            mapping = self._filter_metadata_by_fields(doc.metadata)
+            mapping.update({self._content_field: json.dumps(doc.page_content)})
+
+            # Remove existing key in Redis to avoid reusing the doc ID.
+            self._redis.delete(doc_ids[i])
+            self._redis.hset(doc_ids[i], mapping=mapping)
+
+    def _filter_metadata_by_fields(self, metadata: Optional[dict]) -> dict:
+        """Filter metadata fields to be stored in Redis.
+
+        Args:
+            metadata: The metadata field of a Document object.
+
+        Returns:
+            dict: A subset dict of the metadata that only contains the fields
+                specified in the initialization of the saver. The value of each
+                key is serialized by JSON.
+        """
+        if not metadata:
+            return {}
+        filtered_fields = (
+            self._metadata_fields & metadata.keys()
+            if self._metadata_fields
+            else metadata.keys()
+        )
+        filtered_metadata = {k: json.dumps(metadata[k]) for k in filtered_fields}
```

Comment:
I think the consensus is to save the top level field as-is and only encode the field if it is a "composite" structure. please check out "add_texts" in the vector store integration for references/alignment